### PR TITLE
Feature/fight simulator

### PR DESCRIPTION
**Pull request by Taloreff-hls**
# Fight Simulator Implementation

This PR introduces the core mechanics of the fight simulator, including both the combat system and the catching mechanism.

### Features

- **Pokemon Selection:**
  - Players can choose which of their caught Pokemon will fight in the battle.
  - Players cannot fight against Pokemon they already own.

- **Turn-Based Combat:**
  - The Pokemon with higher speed attacks first.
  - Each turn, one Pokemon attacks while the other defends.
  - Damage is calculated based on the difference between the attacker’s "Attack" stat and the defender’s "Defense" stat.

- **Catch Mechanism:**
  - Players can attempt to catch the opponent's Pokemon during a battle.
  - Catch rate increases as the opponent’s HP decreases.
  - A maximum number of catch attempts is allowed. Failing to catch within the limit will cause the opponent to flee.
  
  **Example:**
  - If the maximum number of attempts is 3:
    - Catching on the second try: The Pokemon is caught and added to the player’s collection.
    - Catching on the third try: The Pokemon is caught and added.
    - Failing on the third try: The Pokemon flees and the fight ends.

### End Conditions

- The fight ends when:
  - A Pokemon’s HP reaches zero.
  - The opponent Pokemon is successfully caught.
  - The opponent Pokemon flees after failing the maximum number of catch attempts.

### CODE REVIEW

File: src/cmps/CatchResultModal.tsx [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  pokemon,
+  onClose,
+}: CatchResultModalProps) => {
+  const [hasResult, setHasResult] = useState(false);
+
+  useEffect(() => {
+    if (!isCaught && hasResult) {
+      const timer = setTimeout(() => {
+        onClose();
+      }, 2000);
```
Comment (ronc1812) [inline src/cmps/CatchResultModal.tsx]: use env var for the timeout number

File: src/cmps/CatchResultModal.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+      return () => clearTimeout(timer);
+    }
+  }, [isCaught, onClose, hasResult]);
+
+  useEffect(() => {
+    const resultTimer = setTimeout(() => {
+      setHasResult(true);
+    }, 3000);
```
Comment (ronc1812) [inline src/cmps/CatchResultModal.tsx]: use env var for the timeout number

File: src/cmps/CatchResultModal.tsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                </PokemonStat>
+                <PokemonStat>
+                  <StatCategory>Category</StatCategory>
+                  <Ability>{pokemon.type[0]}</Ability>
+                </PokemonStat>
+                <PokemonStat>
+                  <StatCategory>Abilities</StatCategory>
+                  <Ability>{pokemon.profile.ability[0][0]}</Ability>
+                </PokemonStat>
+              </PokemonStats>
```
Comment (ronc1812) [inline src/cmps/CatchResultModal.tsx]: this code structure repeat itself i would try to create a new data type and use map to render the items

File: src/cmps/CatchResultModal.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import { Pokemon } from "../interfaces/Pokemon";
+import Typography from "../styles/Typography";
+import colors from "../assets/constants/colors";
+import GenericButton from "../genericCmps/button/GenericButton";
+import { Link } from "react-router-dom";
+import { ResultInnerContainer } from "../styles/StyledFightResultModal";
+import { TypographyType } from "../enums/TypographyEnum";
+
+interface CatchResultModalProps {
+  isCaught: boolean | null;
```
Comment (ronc1812) [inline src/cmps/CatchResultModal.tsx]: why boolean | null, null is falsey

File: src/cmps/CatchResultModal.tsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+              </PokemonStats>
+            </DetailsContainer>
+            <Link to="/">
+              <GenericButton
+                type="primary"
+                size="medium"
+                label="Back to home"
+                fontSize="1.4rem"
+                fontWeight="400"
+                margin={`${SPACING[4]} 0 0 auto`}
```
Comment (ronc1812) [inline src/cmps/CatchResultModal.tsx]: use margin top

File: src/cmps/CatchResultModal.tsx [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                  <StatCategory>Height</StatCategory>
+                  <Ability>{pokemon.profile.height}</Ability>
+                </PokemonStat>
+                <PokemonStat>
+                  <StatCategory>Weight</StatCategory>
+                  <Ability>{pokemon.profile.weight}</Ability>
+                </PokemonStat>
+                <PokemonStat>
+                  <StatCategory>Category</StatCategory>
+                  <Ability>{pokemon.type[0]}</Ability>
```
Comment (ronc1812) [inline src/cmps/CatchResultModal.tsx]: is there a possibility for more than one type?

File: src/cmps/FightArena.tsx [hunk 7]
```diff
@@ -1,63 +1,129 @@
...
+    isFightClicked,
+    currentTurn,
+    battleResult,
+    userHit,
+    opponentHit,
+    showCatchModal,
+    catchResult,
+    setBattleResult,
+    setShowCatchModal,
+  } = battleState;
```
Comment (ronc1812) [inline src/cmps/FightArena.tsx]: use destructuring straight from the hook

File: src/cmps/FightArena.tsx [hunk 8]
```diff
@@ -1,63 +1,129 @@
...
+        <FightResultModal
+          result={battleResult}
+          onClose={() => setBattleResult(null)}
+          playerPokemon={selectedPokemon}
+          opponentPokemon={opponentPokemon}
+        />
+      )}
+      {showCatchModal && opponentPokemon && (
+        <CatchResultModal
+          isCaught={catchResult === true}
```
Comment (ronc1812) [inline src/cmps/FightArena.tsx]: if catchResult is boolean then its unnecessary to check if equals true

File: src/cmps/FightArena.tsx [hunk 9]
```diff
@@ -1,63 +1,129 @@
...
-          Fight
-        </ArenaButton>
-      ) : (
-        <StyledBtnsContainer>
+    <>
+      <StyledFightArena>
+        <GamingCard
+          pokemon={selectedPokemon}
+          hp={userHP}
+          isUser={true}
```
Comment (ronc1812) [inline src/cmps/FightArena.tsx]: you can just pass isUser with out ={true}

File: src/cmps/FightArena.tsx [hunk 10]
```diff
@@ -1,63 +1,129 @@
...
+        )}
+        {opponentPokemon && (
+          <GamingCard
+            pokemon={opponentPokemon}
+            hp={opponentHP}
+            isUser={false}
+            activeTurn={currentTurn === "opponent"}
+            hit={opponentHit}
+            isFightClicked={isFightClicked}
+          />
```
Comment (ronc1812) [inline src/cmps/FightArena.tsx]: use renderOpponent()

File: src/cmps/FightArena.tsx [hunk 11]
```diff
@@ -1,63 +1,129 @@
...
+        )}
+      </StyledFightArena>
+      {battleResult && (
+        <FightResultModal
+          result={battleResult}
+          onClose={() => setBattleResult(null)}
+          playerPokemon={selectedPokemon}
+          opponentPokemon={opponentPokemon}
+        />
+      )}
```
Comment (ronc1812) [inline src/cmps/FightArena.tsx]: use battleResult()

File: src/cmps/FightArena.tsx [hunk 12]
```diff
@@ -1,63 +1,129 @@
...
+      {showCatchModal && opponentPokemon && (
+        <CatchResultModal
```
Comment (ronc1812) [inline src/cmps/FightArena.tsx]: use render catchModal()

File: src/cmps/FightArena.tsx [hunk 13]
```diff
@@ -1,63 +1,129 @@
...
+              type={TypographyType.SubheadingXl}
+              color={colors.neutrals.white}
+              width="237px"
+              height="90px"
+              onClick={handleCatch}
+            >
+              Catch
+            </ArenaButton>
+          </StyledBtnsContainer>
+        )}
```
Comment (ronc1812) [inline src/cmps/FightArena.tsx]: renderButtons()

File: src/cmps/FightResultModal.tsx [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+      detailText = `Your ${playerPokemon.name.english} was defeated by ${opponentPokemon?.name.english}.`;
+      resultPokemon = opponentPokemon;
+      break;
+    case "uncaught":
+      resultText = `${opponentPokemon?.name.english} has fled!`;
+      detailText = `${opponentPokemon?.name.english} has fled after your last catch attempt.`;
+      resultPokemon = opponentPokemon;
+      break;
+    default:
+  }
```
Comment (ronc1812) [inline src/cmps/FightResultModal.tsx]: extract to function/hook, use type/enum for result.

File: src/cmps/FightResultModal.tsx [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+          <PokemonContainer>
+            <PokemonImage
+              src={resultPokemon?.image.hires}
+              alt={resultPokemon?.name.english}
+            />
+          </PokemonContainer>
+          <Typography
+            type={TypographyType.HeadingMd}
+            color={colors.greys[700]}
+            margin={`0 0 ${SPACING[4]} 0`}
```
Comment (ronc1812) [inline src/cmps/FightResultModal.tsx]: use marginBottom

File: src/cmps/FightResultModal.tsx [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+            fontWeight={600}
+          >
+            {titleDetailText}
+          </Typography>
+          <Typography
+            type={TypographyType.Body}
+            color={colors.greys[700]}
+            margin={`0 0 ${SPACING[4]} 0`}
```
Comment (ronc1812) [inline src/cmps/FightResultModal.tsx]: use marginBottom

File: src/cmps/FightResultModal.tsx [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+          </Typography>
+
+          <Link to="/">
+            <GenericButton
+              type="primary"
+              size="medium"
+              label="Back to Home"
+              fontSize="1.4rem"
+              fontWeight="400"
+              margin={`${SPACING[4]} 0 0`}
```
Comment (ronc1812) [inline src/cmps/FightResultModal.tsx]: use marginTop

File: src/cmps/GamingCard.tsx [hunk 18]
```diff
@@ -1,39 +1,84 @@
...
+            setIsAnimating(false);
+          }
+          return newHP;
+        });
+      }, 30);
+
+      return () => clearInterval(interval);
+    } else if (hit === 0 && activeTurn && isFightClicked) {
+      setIsAnimating(true);
+      setTimeout(() => setIsAnimating(false), 1000);
```
Comment (ronc1812) [inline src/cmps/GamingCard.tsx]: use env var for the timeout time

File: src/cmps/GamingCard.tsx [hunk 19]
```diff
@@ -60,16 +105,16 @@ const GamingCard = ({ pokemon }: GamingCardProps) => {
         type={TypographyType.SubheadingMd}
         color={colors.neutrals[200]}
         aligntext="center"
-      >{`#${pokemon?.id.toString().padStart(4, "0")}`}</Typography>
+      >{`#${pokemon.id.toString().padStart(4, "0")}`}</Typography>
```
Comment (ronc1812) [inline src/cmps/GamingCard.tsx]: save to memo

File: src/cmps/GamingCard.tsx [hunk 20]
```diff
@@ -1,39 +1,84 @@
...
+      const interval = setInterval(() => {
+        setAnimatedHP((prevAnimatedHP) => {
+          const newHP = Math.max(prevAnimatedHP - 1, hp);
+          if (newHP === hp) {
+            clearInterval(interval);
+            setIsAnimating(false);
+          }
+          return newHP;
+        });
+      }, 30);
```
Comment (ronc1812) [inline src/cmps/GamingCard.tsx:54]: can you use useMemo instead of useEffect?

File: src/cmps/Header.tsx [hunk 21]
```diff
@@ -46,9 +46,9 @@ const Header = ({ setSelectedCtg, selectedCtg }: HeaderProps) => {
       <Link to="/fight-arena">
         <GenericButton
           type="primary"
-          size="large"
+          size="medium"
```
Comment (ronc1812) [inline src/cmps/Header.tsx]: enum

File: src/cmps/PokemonGrid.tsx [hunk 22]
```diff
@@ -58,13 +60,22 @@ const PokemonGrid = ({ pokemons }: PokemonGridProps) => {
...
+              onClick={() => handleCardClick(pokemon)}
+            />
+          ))
+        ) : (
+          <EmptyPokemonsContainer>
+            <EmptyPokemons>
+              <img src={emptyPokemons} alt="no pokemons" />
+              No Pokemons were found.
+            </EmptyPokemons>
+          </EmptyPokemonsContainer>
```
Comment (ronc1812) [inline src/cmps/PokemonGrid.tsx]: create an emptyState wrapper that recieve children and if the children are null|undefiend render the empty state, else the children

File: src/cmps/PokemonTable.tsx [hunk 23]
```diff
@@ -28,13 +29,18 @@ import { NO_VALUE } from "../assets/constants/placeholders";
...
 
-const PokemonTable = ({ pokemons }: PokemonTableProps) => {
+const PokemonTable = ({ pokemons, selectedCtg }: PokemonTableProps) => {
   const [page, setPage] = useState(0);
   const [rowsPerPage, setRowsPerPage] = useState(10);
   const [selectedPokemon, setSelectedPokemon] = useState<Pokemon | null>(null);
 
+  useEffect(() => {
+    setPage(0);
+  }, [selectedCtg]);
```
Comment (ronc1812) [inline src/cmps/PokemonTable.tsx:42]: every change of ctg you need to reset the page?

**Review by ronc1812 — CHANGES_REQUESTED** [review]

File: src/cmps/GamingCard.tsx (hunk 20 above)
Comment (Taloreff-hls) [reply src/cmps/GamingCard.tsx:54]: i think useEffect is more suitable to this situation no? its a side effect and and the decreased HP is always a different calculation

File: src/cmps/PokemonTable.tsx (hunk 23 above)
Comment (Taloreff-hls) [reply src/cmps/PokemonTable.tsx:42]: for this case:
user is in All Pokemons tab, using pagination till page number 10, switch to My Pokemons tab, but user only have 3 pokemons, so i need to set the pagination to page number 0

**Review by DinFridman — APPROVED** [review]

**Pull request merged**